src/lavinmq/mqtt/retain_store.cr

```diff
@@ -0,0 +1,144 @@
+require "./topic_tree"
+require "digest/sha256"
+
+module LavinMQ
+  module MQTT
+    class RetainStore
+      Log = LavinMQ::Log.for("retainstore")
+
+      MESSAGE_FILE_SUFFIX = ".msg"
+      INDEX_FILE_NAME     = "index"
+
+      alias IndexTree = TopicTree(String)
+
+      def initialize(@dir : String, @index = IndexTree.new)
+        Dir.mkdir_p @dir
+        @index_file = File.new(File.join(@dir, INDEX_FILE_NAME), "a+")
+        @lock = Mutex.new
+        @lock.synchronize do
+          if @index.empty?
+            restore_index(@index, @index_file)
+            write_index
+            @index_file = File.new(File.join(@dir, INDEX_FILE_NAME), "a+")
+          end
+        end
+      end
+
+      def close
+        @lock.synchronize do
+          write_index
+        end
+      end
+
+      private def restore_index(index : IndexTree, index_file : ::IO)
+        Log.info { "restoring index" }
+        dir = @dir
+        msg_count = 0
+        msg_file_segments = Set(String).new(
+          Dir[Path[dir, "*#{MESSAGE_FILE_SUFFIX}"]].compact_map do |fname|
+            File.basename(fname)
+          end
+        )
+
+        while topic = index_file.gets
+          msg_file_name = make_file_name(topic)
+          unless msg_file_segments.delete(msg_file_name)
+            Log.warn { "msg file for topic #{topic} missing, dropping from index" }
+            next
+          end
+          index.insert(topic, msg_file_name)
+          Log.debug { "restored #{topic}" }
+          msg_count += 1
+        end
+
+        # TODO: Device what's the truth: index file or msgs file. Mybe drop the index file and rebuild
+        # index from msg files?
+        unless msg_file_segments.empty?
+          Log.warn { "unreferenced messages: #{msg_file_segments.join(",")}" }
+        end
+        # TODO: delete unreferenced messages?
+        Log.info { "restoring index done, msg_count = #{msg_count}" }
+      end
+
+      def retain(topic : String, body_io : ::IO, size : UInt64) : Nil
+        @lock.synchronize do
+          Log.debug { "retain topic=#{topic} body.bytesize=#{size}" }
+          # An empty message with retain flag means clear the topic from retained messages
+          if size.zero?
+            delete_from_index(topic)
+            return
+          end
+
+          unless msg_file_name = @index[topic]?
+            msg_file_name = make_file_name(topic)
+            add_to_index(topic, msg_file_name)
+          end
+
+          File.open(File.join(@dir, msg_file_name), "w+") do |f|
```

Comment:
You open the file each time?

Reply:
And over writing it? Without doing the "write then rename"? Can cause half written files now.